### PR TITLE
hls-pragmas-plugin: Reduce noisy completions

### PR DESCRIPTION
**Pull request by akshaymankar**
#3617

### CODE REVIEW

**Review by fendor — APPROVED** [review]
LGTM

**Comment by michaelpj** [comment]
Could we get some comments? Sounds like it was somewhat non-obvious what to do? Not clear to me what the code does either...

**Comment by akshaymankar** [comment]
Thanks @michaelpj! I added comments, which helped me simplify things :)

File: plugins/hls-pragmas-plugin/src/Ide/Plugin/Pragmas.hs [hunk 1]
```diff
@@ -181,23 +182,41 @@ completion _ide _ complParams = do
...
+                      --    'LANGUAGE' pragma.
+                      -- 3. The user has not typed anything yet.
+                      "import" `T.isPrefixOf` line || not (T.null module_) || T.null word
+                    = []
                     | otherwise
-                    = J.List $ [ mkPragmaCompl (prefix <> a <> suffix) b c
-                                | (a, b, c, _) <- validPragmas, Fuzzy.test word b]
+                    = [ mkPragmaCompl (prefix <> pragmaTemplate <> suffix) matcher detail
+                      | (pragmaTemplate, matcher, detail, appearWhere) <- validPragmas
+                      , -- Only suggest a pragma that need its own line if the whole line
```
Comment (fendor) [inline plugins/hls-pragmas-plugin/src/Ide/Plugin/Pragmas.hs:209]:
```suggestion
                      , -- Only suggest a pragma that needs its own line if the whole line
```

**Review by fendor — APPROVED** [review]

File: plugins/hls-pragmas-plugin/test/Main.hs [hunk 2]
```diff
@@ -127,29 +130,80 @@ completionSnippetTests :: TestTree
...
+        Nothing -> compls @?= []
+        Just undesiredLabel -> do
+            case find (\c -> c ^. L.label == undesiredLabel) compls of
+                Just c -> assertFailure $
+                    "Did not expect a completion with label=" <> T.unpack undesiredLabel
+                    <> ", got completion: "<> show c
+                Nothing -> pure ()
+
+disableGhcideCompletions :: Value
+disableGhcideCompletions = object [ "haskell" .= object ["plugin" .= object [ "ghcide-completions" .= object ["globalOn" .= False]]] ]
```
[the remark this answers is not in the document]
Comment (fendor) [reply plugins/hls-pragmas-plugin/test/Main.hs:205]: Created one: #3653

**Comment by michaelpj** [comment]
Has a conflict now

**Pull request merged**